src/core/image.js

```diff
@@ -535,6 +535,128 @@ var PDFImage = (function PDFImageClosure() {
+     * @return {Boolean}
+     */
+    shallResizeImage: function PDFImage_shallResizeImage(comps, bpc) {
+      return (comps === 1 && (bpc === 1 || bpc === 8));
+    },
+
+    /**
+     * Resize black white image data
+     * @param {Uint8Array} imgData image data.
+     * @param {Number} scaleBits
+     * @author Ramsoft
+     */
+    resizeBWPixels: function PDFImage_resizeBWPixels(imgData, scaleBits) {
+      var newWidth = imgData.width >> scaleBits;
+      var newHeight = imgData.height >> scaleBits;
+
+      //we want to output 1 byte per pixel
+      var newRowBytes = (newWidth + 7) >> 3;
+      var originalRowBytes = (imgData.width + 7) >> 3;
+      var step = 1 << scaleBits;
+      var numBytes = newHeight * newRowBytes;
+      var pixelArrayOutput = new Uint8Array(numBytes);
+
+      for (var i = 0, y2 = 0; i < newHeight; i++, y2 += step) {
+        var originalRowStart = y2 * originalRowBytes;
+        var newRowStart = i * newRowBytes;
+
+        for (var j = 0, x2 = 0; j < newWidth; j++, x2 += step) {
+          // we want original value of pixel [x2, y2] value in original image
+          // to populate pixel [i, j] in new image
+
+          var originalColByteStart = x2 >> 3;
+          var originalColBitMask = 1 << (7 - (x2 & 7));
+          // most signifcant bit is first pixel due to Little Endian
+
+          var originalColByte =
+            imgData.data[originalRowStart + originalColByteStart];
+          var pixelValue = originalColByte & originalColBitMask;
+
+          if (pixelValue > 0) {
+            // most signifcant bit is first pixel due to Little Endian
+            var newColByteStart = j >> 3;
+            var newColBitMask = 1 << (7 - (j & 7));
+
+            var newColByte = pixelArrayOutput[newRowStart + newColByteStart];
+
+            newColByte = newColByte | newColBitMask;
+            // set pixel bit to 1
+            pixelArrayOutput[newRowStart + newColByteStart] = newColByte;
+          }
+        }
+      }
+
+      imgData.data = pixelArrayOutput;
+      imgData.width = newWidth;
+      imgData.height = newHeight;
+    },
+
+    /**
+     * Resize large resolution PDFS as to improve rendering time
+     * @param  {Uint8Array} imgData image data.
+     * @param  {Number} comps Number of color components.
+     * @param  {Number} bpc Number of bits per component.
+     */
+    resizeGrayPixels: function PDFImage_resizeGrayPixels(imgData, comps, bpc) {
+      var scaleBits;
+      var h1 = imgData.height;
+      var w1 = imgData.width;
+
+      // Reduce the pixel dimensions to around 2000.
+      // to improve performance for large grayscale bitmaps > 10K width 
+      // or height
+      // If Width or Height > 15K, scale down the size to 10% of original size.
+      // If Width or Height > 10K, scale down the size to 15% of original size.
+      // If Width or Height > 5K, scale down the size to 25% of original size.
+      // Otherwise, use the original size.
```

Comment:
The documentation does not reflect what your code does. If you want to achieve a target size of 2000 to 4000 pixel, you could loop until this target size is reached. This would reduce the need for many magic numbers.
```
var MAX_SIZE = 4000;
var scaleBits = 0;
while (((w1 >> scaleBits) > MAX_SIZE) || ((h1 >> scaleBits) > MAX_SIZE)) {
	scaleBits++; 
}
```